### PR TITLE
Add GitHub proxy

### PR DESCRIPTION
**Pull request by bloodearnest**
Restricts to clone/pull/push from `opensafely` and `opensafely-core`
orgs only.

Push will ideally be disabled in future, but is still needed currently.

There are two sets of tests: CI tests, which must run w/o TLS and DNS
for the proxies, so are basic http calls. These tests are run as landing
tests. They use curl, as spoofing DNS in python is non-trivial.

The second set of tests full-tests.sh, which will only run against
production, as it needs full TLS/DNS set up. But it tests the actual
docker and git clients against the proxy.

Also a few clean ups of the base docker image and ghcr.io proxy

### CODE REVIEW

File: github.com.conf.template [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    server_name github-proxy.opensafely.org;
+    root /var/www/html/;
+    listen ${PORT};
+
+    # The default nginx request body size is 1M, we make it explicit and bump
+    # to support current use cases
+    client_max_body_size 2M;
+
+    # only opensafely and opensafley-core orgs can be accessed
```
Comment (benbc) [inline github.com.conf.template:21]: Why is opensafely-core needed here?

File: github.com.conf.template [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    location ~ /(opensafely|opensafely-core)/.*/(info|git-upload-pack|git-receive-pack) {
```
Comment (benbc) [inline github.com.conf.template:22]: I'd be inclined to pin the regex to the start of the location like `^/(opensafely|opensafely-core)/...`.

Comment (benbc) [inline github.com.conf.template:22]: The regex here doesn't seem to be consistent with the use of `git-upload-pack` etc in the query fragment -- it looks like they're part of the path here, but that's not right, is it?

**Review by benbc — COMMENTED** [review]
Looks good, Simon. I mostly just have questions due to not understanding stuff properly yet.

File: ghcr.io.conf.template [hunk 3]
```diff
@@ -41,23 +44,14 @@ server {
     location /token {
         limit_except GET { deny all; }
         proxy_pass https://ghcr.io;
+        proxy_redirect default;
     }
 
     # opensafely images only, and handle redirects
-    location /v2/opensafely/ {
```
[the remark this answers is not in the document]
Comment (bloodearnest) [reply ghcr.io.conf.template:47]: Not anymore - the docker images are publish under the `opensafely-core` or now, and I relealised that allowing opensafely images was actually a hole, as anyone in the `opensafely` org can publish an arbitrary image, so I cleaned it up

File: ghcr.io.conf.template [hunk 4]
```diff
@@ -41,23 +44,14 @@ server {
...
-        # ghcr.io redirects to an S3 bucket, which is not accessible from
-        # backends. So handle redirects to S3 here in the proxy rather than
-        # passing back to the client
-        proxy_intercept_errors on;
-        error_page 301 302 307 = @handle_redirect;
-    }
-
-    # opensafely-core images only, and handle redirects
-    location /v2/opensafely-core/ {
+    location ~ /v2/opensafely-core/ {
```
[the remark this answers is not in the document]
Comment (bloodearnest) [reply ghcr.io.conf.template:51]: So, all the repos that publish the images are in the `opensafely-core` org, as it's a core part of the infra. Only the datalab tech team have the rights to publish these images. 

Basically, the images used to be published under the opensafely org, they are now published under the opensafely-core org, and this is my driveby cleaning it up. 

Ideally this should have been a separate PR, sorry for the confusion.

[the remark this answers is not in the document]
Comment (bloodearnest) [reply ghcr.io.conf.template:51]: Ah, that's my error - I was playing with `location ~ /v2/(opensafely|opensafely-core)/, but decided to remove `opensafely` only bit, forgot to revert the regex, thanks!

File: github.com.conf.template (hunk 1 above)
Comment (bloodearnest) [reply github.com.conf.template:21]: Good question. All the study code is in opensafely, but the infra code, like job-runner is in opensafely-core. We currently deploy new job-runner changes by pulling updates from githhub - hence why both are needed. We used to use docker images to deploy job-runner, but the windows-on-docker issues we had mean we deploy via a `git pull` now.

File: github.com.conf.template (hunk 2 above)
Comment (bloodearnest) [reply github.com.conf.template:22]: Good call, will do.

Comment (bloodearnest) [reply github.com.conf.template:22]: It is right! At least, if I've understood it correctly. It certainly seems to work. 

tl;dr: it works by appending `/somepath` to the end of the base repo url. So depending on your base repo url, it could be an actual path component (like github does) or part of a query string.

Basically, git's http protocol is not very http like. Search for $GIT_URL here for examples: https://git-scm.com/docs/http-protocol

File: README.md [hunk 5]
```diff
@@ -1,54 +1,45 @@
...
+domains:
+ 
+ * github-proxy.opensafely.org: this provides access to *only* opensafely
+   repositories hosted on https://github.com, and not other repositories.
 
-So this proxy provides a single point of read-only access to only the
-OpenSAFELY images on ghcr.io, by only proxying GET requests to /v2/opensafely/
-urls.
+ * docker-proxy.opensafely.org: this provides read only access to docker images
+   published by specific organistions on https://ghcr.io, the Github Container
```
Comment (madwort) [inline README.md:14]: nit: I think `specific Github organisations` would be better here (which means some of the several GH orgs our team has), `specific organisations` implies to me real-world orgs (like, Canonical, DL, ...)

File: ci-tests.sh [hunk 6]
```diff
@@ -0,0 +1,116 @@
+#!/bin/bash
+set -euo pipefail
+# In CI, we do not have TLS/DNS set up. Docker requires TLS, so we cannot do
+# proper functional tests until post deployment. And most tools require DNS in
+# order to hit the right vhost in nginx.
+#
+# So we use curl's --resolve opti0on to query the base APIs to check we get the
```
Comment (madwort) [inline ci-tests.sh:7]: typo `opti0on`, you hax0r

File: ci-tests.sh [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    if test "$(cat "$body")" == "$expected"; then
+        ok "$url returned expected body"
+    else
+        fail "$url returned unexpected body:"
+    fi
+}
+
+assert-header() {
+    if test "$last_test_failed" = "1"; then
+        echo "SKIP assert body"
```
Comment (madwort) [inline ci-tests.sh:71]: possible typo: should this be `SKIP assert header`?

File: ci-tests.sh [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+assert-body() {
+    if test "$last_test_failed" = "1"; then
+        echo "SKIP assert body"
+        return
+    fi
+    local expected=$1
+    if test "$(cat "$body")" == "$expected"; then
+        ok "$url returned expected body"
+    else
+        fail "$url returned unexpected body:"
```
Comment (madwort) [inline ci-tests.sh:65]: should this have a `$body` at the end to output the bad body?

File: README.md (hunk 5 above)
Comment (bloodearnest) [reply README.md:14]: good call, will change

File: ci-tests.sh (hunk 6 above)
Comment (bloodearnest) [reply ci-tests.sh:7]: I is 1337

File: ci-tests.sh (hunk 7 above)
Comment (bloodearnest) [reply ci-tests.sh:71]: yes, will fix

File: ci-tests.sh (hunk 8 above)
Comment (bloodearnest) [reply ci-tests.sh:65]: the fail() function outputs the body by default?

Comment (bloodearnest) [reply ci-tests.sh:65]: ah, the trailing :, will fix

File: ghcr.io.conf.template [hunk 9]
```diff
@@ -1,36 +1,39 @@
...
+# these need to be global, sadly
 proxy_http_version 1.1; # keep alives to upstream
 proxy_buffering off;    # start streaming response immediately
-#error_log /var/log/nginx/error-debug.log debug;
 
 server {
 
-    # be identified by either domain
-    server_name ${PROXY_DOMAIN} ${PROXY_ORIGIN};
+    # note: embdatalab domain is for when we are being proxied by cloudflare
```
Comment (madwort) [inline ghcr.io.conf.template:22]: Not sure I quite follow - do you mean "the ebmdatalab domain allows you direct access to the server when the primary domain is being proxied by CloudFlare"?

Comment (bloodearnest) [reply ghcr.io.conf.template:22]: Kinda - it's so nginx picks the correct vhost when cloudflare is proxying `docker-proxy.opensafely.org `to the origin domain of `docker-proxy.ebmdatalab.net` (as we are currently). So the vhost can respond to either domain. When EMIS add our new floating IP, plan is to disable proxying, and have nginx host the `docker-proxy.opensafely.org` domain directly. Hence why it has both.

Comment (madwort) [reply ghcr.io.conf.template:22]: ah ok, let's tidy when EMIS add this ip then

File: github.com.conf.template [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    server_name github-proxy.opensafely.org;
+    root /var/www/html/;
+    listen ${PORT};
+
+    # The default nginx request body size is 1M, we make it explicit and bump
+    # to support current use cases
+    client_max_body_size 2M;
+
+    # only opensafely and opensafely-core orgs can be accessed
+    location ~ ^/(opensafely|opensafely-core)/.*/(info|git-upload-pack|git-receive-pack) {
```
Comment (madwort) [inline github.com.conf.template:22]: hmmm, is there scope to spoof this regex? we want to accept urls such as `github-proxy.opensafely.org/opensafely-core/job-runner/info/refs?service=git-upload-pack` but what about urls such as `github-proxy.opensafely.org/opensafely-core/job-runner/info/refs?service=whatever-you-want&haxor=git-upload-pack`?

Comment (bloodearnest) [reply github.com.conf.template]: I have attempted to improve the situation

File: github.com.conf.template [hunk 11]
```diff
@@ -19,11 +19,19 @@ server {
...
     # only opensafely and opensafely-core orgs can be accessed
-    location ~ ^/(opensafely|opensafely-core)/.*/(info|git-upload-pack|git-receive-pack) {
+    # location regex cannot match on query parameters
+    location ~ ^/(opensafely|opensafely-core)/[^/]+/info/refs {
+        if ($args !~ "service=git-(upload|receive)-pack") { return 404; }
         proxy_pass https://github.com;
         proxy_redirect default;
     }
 
+    location ~ ^/(opensafely|opensafely-core)/[^/]+/git-(upload|receive)-pack {
```
Comment (madwort) [inline github.com.conf.template:29]: 👍 `[^/]` much nicer than `.*`

**Review by madwort — APPROVED** [review]
